src/libmongoc/tests/test-mongoc-sdam.c

```diff
@@ -215,6 +216,250 @@ test_sdam_cb (bson_t *test)
+   coll_name = bson_lookup_utf8 (ctx->config->scenario, "collection_name");
+   ctx->client = client;
+   ctx->db = mongoc_client_get_database (client, db_name);
+   ctx->collection = mongoc_database_get_collection (ctx->db, coll_name);
+}
+
+static void
+sdam_json_test_ctx_cleanup (json_test_ctx_t *ctx)
+{
+   mongoc_collection_destroy (ctx->collection);
+   mongoc_database_destroy (ctx->db);
+   bson_destroy (&ctx->lsids[0]);
+   bson_destroy (&ctx->lsids[1]);
+   bson_destroy (&ctx->events);
+   mongoc_uri_destroy (ctx->test_framework_uri);
+   bson_destroy (ctx->sent_lsids[0]);
+   bson_destroy (ctx->sent_lsids[1]);
+   bson_mutex_destroy (&ctx->mutex);
+}
+
+static bool
+sdam_integration_operation_cb (json_test_ctx_t *ctx,
+                               const bson_t *test,
+                               const bson_t *operation)
+{
+   bson_t reply;
+   bool res;
+
+   res =
+      json_test_operation (ctx, test, operation, ctx->collection, NULL, &reply);
+
+   bson_destroy (&reply);
+
+   return res;
+}
+
+/* Try to get a completely clean slate by disabling failpoints on all servers.
+ */
+static void
+deactivate_failpoints_on_all_servers (mongoc_client_t *client)
+{
+   int i;
+   uint32_t server_id;
+   mongoc_set_t *servers;
+   bson_t cmd;
+   bson_error_t error;
+
+   bson_init (&cmd);
+   BCON_APPEND (&cmd, "configureFailPoint", "failCommand", "mode", "off");
+
+   servers = client->topology->description.servers;
+
+   for (i = 0; i < servers->items_len; i++) {
+      bool ret;
+
+      server_id = servers->items[i].id;
+      ret = mongoc_client_command_simple_with_server_id (client,
+                                                         "admin",
+                                                         &cmd,
+                                                         NULL /* read prefs */,
+                                                         server_id,
+                                                         NULL /* reply */,
+                                                         &error);
+      if (!ret) {
+         MONGOC_DEBUG ("error disabling failpoint: %s", error.message);
+      }
+   }
+
+   bson_destroy (&cmd);
+}
+
+static void
+run_one_integration_test (json_test_config_t *config, bson_t *test)
+{
+   json_test_ctx_t ctx;
+   json_test_ctx_t thread_ctx[2];
+   bson_error_t error;
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *setup_client;
+   mongoc_client_t *test_client;
+   mongoc_client_t *thread_clients[2];
+   const char *db_name;
+   const char *coll_name;
+   mongoc_uri_t *uri;
+
+   MONGOC_DEBUG ("running test: %s", bson_lookup_utf8 (test, "description"));
+
+   uri = test_framework_get_uri ();
+   if (bson_has_field (test, "clientOptions")) {
+      bson_t client_opts;
+
+      bson_lookup_doc (test, "clientOptions", &client_opts);
+      set_uri_opts_from_bson (uri, &client_opts);
+   }
+   mongoc_uri_set_option_as_bool (uri, MONGOC_URI_DIRECTCONNECTION, false);
```

Comment:
Is this necessary? Doesn't the connection string returned by `test_framework_get_uri` already ensure that replica sets also specify a `replicaSet` option that would ensure discovery behavior?

Or does this perhaps have implications for connection strings with only a single host? In that case, I expect the topology would still end up being sharded due to `isdbgrid` being found in the `isMaster` response.

Reply:
Noted.

IIRC, libmongoc makes decisions about feature support by server type (in addition to wire version, of course) and never the topology type. That's why things like retryable writes and transactions would still work when directly connected to a replica set primary or mongos node.